### PR TITLE
Improve warnings for constraints and materialization types

### LINKED ISSUE #7335: [CT-2402] [Feature] Improve warnings/errors for constraints <> materialization types
### Views

If your model is materialized as a `view` and has `constraints` defined, those constraints are not supported & will be ignored, on ALL data platforms. We should raise a warning saying this, similar to the `warn_unsupported` functionality added in #7250.

**Update:** This might require calling a different macro within `create_view_as`, since all constraint parsing/rendering is currently happening within `create_table_as` &rarr; `get_table_columns_and_constraints`.

---

### Ephemeral models
```
Parsing Error
  Constraint validation failed for: (models/sometable.sql)
  Only table, view, and incremental materializations are supported for constraints, but found 'ephemeral'
```
Let's just update the language here: "constraint validation" &rarr; "contract enforcement"

https://github.com/dbt-labs/dbt-core/blob/87e25e8692197c093e9be2c0df5fe38dea48eec5/core/dbt/parser/schemas.py#L969-L988

### PR DESCRIPTION
**Pull request by emmyoop**
resolves #7335 

### Checklist

- [ ] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [ ] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [ ] I have run this code in development and it appears to resolve the stated issue
- [ ] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [ ] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: tests/functional/configs/test_contract_configs.py [hunk 1]
```diff
@@ -367,22 +396,49 @@ class TestModelLevelConstraintsErrorMessages:
...
+        )
+        assert expected_materialization_warning in str(log_output)
+
+        # change to not show warnings, message should not be in logs
+        models_dir = os.path.join(project.project_root, "models")
+        write_file(model_schema_ignore_unsupported_yml, models_dir, "constraints_schema.yml")
+        _, log_output = run_dbt_and_capture(["run"])
+
+        expected_materialization_warning = (
+            "Constraint types are not supported for ephemeral materializations."
```
Comment (MichelleArk) [inline tests/functional/configs/test_contract_configs.py:437]: should this be 'view' instead of 'ephemeral'?

File: core/dbt/parser/schemas.py [hunk 2]
```diff
@@ -817,23 +818,39 @@ def patch_constraints(self, node, constraints):
...
+            for column in model_node.columns.values()
+            for constraint in column.constraints
+        ]
+        model_warn_unsupported = [
+            constraint.warn_unsupported for constraint in model_node.constraints
+        ]
+        warn_unsupported = column_warn_unsupported + model_warn_unsupported
+
+        # if any constraint has `warn_unsupported` as True then send the warning
+        if any(warn_unsupported) and model_node.config.materialized not in [
```
Comment (MichelleArk) [inline core/dbt/parser/schemas.py:833]: could be refactored to use [materialization_enforces_constraints](https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/contracts/graph/nodes.py#LL596C9-L596C45): `if any(warn_unsupported) and not model_node.materialization_enforces_constraints`

File: core/dbt/parser/schemas.py [hunk 3]
```diff
@@ -817,23 +818,39 @@ def patch_constraints(self, node, constraints):
...
+                node=model_node,
+            )
+
         errors = []
         if not model_node.columns:
             errors.append(
                 "Constraints must be defined in a `yml` schema configuration file like `schema.yml`."
             )
 
-        if model_node.config.materialized not in ["table", "view", "incremental"]:
```
Comment (MichelleArk) [inline core/dbt/parser/schemas.py:826]: 👍

**Review by MichelleArk — APPROVED** [review]
a couple small suggestions, otherwise LGTM!

File: tests/functional/configs/test_contract_configs.py (hunk 1 above)
Comment (emmyoop) [reply tests/functional/configs/test_contract_configs.py:437]: Great catch!

File: core/dbt/parser/schemas.py (hunk 2 above)
Comment (emmyoop) [reply core/dbt/parser/schemas.py:833]: Yes it 100% should be.

File: tests/functional/configs/test_contract_configs.py (hunk 1 above)
Comment (VersusFacit) [reply tests/functional/configs/test_contract_configs.py:437]: Should this be an adapter zone test?

Comment (emmyoop) [reply tests/functional/configs/test_contract_configs.py]: This specifically is not adapter specific.  There are tests on constraints in the adapter zone already.

**Pull request merged**